zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -189,18 +189,31 @@ impl Chain {
 
     /// Push a contextually valid non-finalized block into this chain as the new tip.
     ///
-    /// If the block is invalid, drop this chain and return an error.
+    /// If the block is invalid, clears the chain, and returns an error.
     ///
     /// Note: a [`ContextuallyValidBlock`] isn't actually contextually valid until
     /// [`update_chain_state_with`] returns success.
     #[instrument(level = "debug", skip(self, block), fields(block = %block.block))]
-    pub fn push(mut self, block: ContextuallyValidBlock) -> Result<Chain, ValidateContextError> {
+    pub fn push(&mut self, block: ContextuallyValidBlock) -> Result<(), ValidateContextError> {
         // update cumulative data members
-        self.update_chain_tip_with(&block)?;
+        if let Err(error) = self.update_chain_tip_with(&block) {
+            // The chain could be in an invalid half-updated state, so clear its data.
+            *self = Chain::new(
+                self.network,
+                sprout::tree::NoteCommitmentTree::default(),
+                sapling::tree::NoteCommitmentTree::default(),
+                orchard::tree::NoteCommitmentTree::default(),
+                HistoryTree::default(),
+                ValueBalance::zero(),
+            );
```

Comment:
This feels weird. I think the previous way having the method take ownership of `self` avoids the risk of misusing a chain after an error. Maybe we can restore that behaviour with a few changes. I played around a bit to see if it was possible, and got this: #3826.

Reply:
You're right, thanks for the simplification!